src/java/com/puppetlabs/puppetserver/pool/JRubyPool.java

```diff
@@ -0,0 +1,89 @@
+package com.puppetlabs.puppetserver.pool;
+
+import java.util.Set;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.CopyOnWriteArraySet;
+import java.util.concurrent.LinkedBlockingDeque;
+import java.util.concurrent.locks.ReentrantReadWriteLock;
+import com.puppetlabs.puppetserver.pool.LockablePool;
+
+public final class JRubyPool<E> implements LockablePool<E> {
+    private final LinkedBlockingDeque<E> live_queue;
+    private final ReentrantReadWriteLock lock = new ReentrantReadWriteLock(true);
+    private final Set<E> registeredElements = new CopyOnWriteArraySet<>();
+
+    public JRubyPool(int size) {
+        live_queue = new LinkedBlockingDeque<>(size);
+    }
+
+    @Override
+    public void register(E e) throws InterruptedException {
+        live_queue.putLast(e);
+        registeredElements.add(e);
+    }
+
+    @Override
+    public E borrowItem() throws InterruptedException, IllegalStateException {
+        if (lock.isWriteLockedByCurrentThread()) {
+          throw new IllegalStateException("The current implementation has a risk of deadlock if you attempt to borrow a JRuby instance while holding the write lock!");
+        }
+        E item = live_queue.takeFirst();
+        lock.readLock().lock();
+        return item;
+    }
+
+    @Override
+    public E borrowItemWithTimeout(long timeout, TimeUnit unit) throws InterruptedException, IllegalStateException {
+        if (lock.isWriteLockedByCurrentThread()) {
+          throw new IllegalStateException("The current implementation has a risk of deadlock if you attempt to borrow a JRuby instance while holding the write lock!");
+        }
+        E item = live_queue.pollFirst(timeout, unit);
+        lock.readLock().lock();
+        return item;
+    }
+
+    @Override
+    public void returnItem(E e) throws InterruptedException {
+        live_queue.putFirst(e);
+        lock.readLock().unlock();
```

Comment:
Typically whenever you're holding a lock it's a good idea to put the `unlock` in a `try`/`finally` block, to be as certain as possible that the lock will be released, because if an exception is thrown somewhere else and the lock isn't released it will cause problems.

In this case it probably won't make much of a difference because there's only one line of code where an exception could happen, and if something went wrong on that line of code, we'd probably already be hosed... but I might be inclined to put the `try`/`finally` in anyway, just because it's generally a good pattern to follow.